### PR TITLE
docstrings + search funcs

### PR DESCRIPTION
**Pull request by KonScanner**
Added docstrings + search function(s).

### CODE REVIEW

**Comment by woctezuma** [comment]
Thank you for the pull request!

I have a few minor remarks:

1) Codacy Static Code Analysis mentions a minor issue with PEP 257 docstring style checker:
> Multi-line docstring summary should start at the first line (D212)

That is the style which is most common. However, it is true that [PEP 257](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings) allows both styles.
> The summary line may be on the same line as the opening quotes or on the next line.

2) I would include the search functions in a separate file, e.g. `src/search_utils.py` due to `import re`, but that is personal preference.

3) I would add a few unit tests for the new functions.

4) I would run [`black` formatter](https://github.com/psf/black) before any merge.

5) I would not include `.gitignore`.

File: src/aggregate_utils.py [hunk 1]
```diff
@@ -1,17 +1,39 @@
-from src.download_utils import get_default_params, download_store_data
+from download_utils import get_default_params, download_store_data
```
Comment (woctezuma) [inline src/aggregate_utils.py:1]: I don't think this change is correct if Python is run from the root of the repository.

File: src/aggregate_utils.py [hunk 2]
```diff
@@ -1,17 +1,39 @@
...
+import re
```
Comment (woctezuma) [inline src/aggregate_utils.py:2]: I would include the search functions in a separate file, e.g. `src/search_utils.py` due to `import re`.

File: src/aggregate_utils.py [hunk 3]
```diff
@@ -24,22 +46,45 @@ def download_all_promos(verbose=True):
     all_promos = []
 
     for cursor in range(0, total_num_promos, step):
-        store_data = download_store_data(cursor=cursor, step=step, verbose=verbose)
+        store_data = download_store_data(
```
Comment (woctezuma) [inline src/aggregate_utils.py:49]: I would run [`black` formatter](https://github.com/psf/black) before any merge.

File: src/aggregate_utils.py [hunk 4]
```diff
@@ -24,22 +46,45 @@ def download_all_promos(verbose=True):
...
+            cursor=cursor, step=step, verbose=verbose)
```
Comment (woctezuma) [inline src/aggregate_utils.py:50]: I would run [`black` formatter](https://github.com/psf/black) before any merge.

File: src/aggregate_utils.py [hunk 5]
```diff
@@ -24,22 +46,45 @@ def download_all_promos(verbose=True):
...
+    """
+    Regex search for title in promotions.
+
+    args: promotions: List[dict], title: str
+
+    Returns: All potential matches.
+    """
+    return [e for e in promotions if title in re.search(title, e["title"], re.IGNORECASE)]
+
+
```
Comment (woctezuma) [inline src/aggregate_utils.py:83]: - I would include the search functions in a separate file, e.g. `src/search_utils.py` due to `import re`.
- I would add a few unit tests for the new functions.

File: src/filter_utils.py [hunk 6]
```diff
@@ -18,15 +32,37 @@ def is_relevant_promo(element, check_upcoming_promos):
...
-    filtered_promos = [e for e in promos if is_relevant_promo(e, check_upcoming_promos)]
+    """
+    Filter promotions.
+
+    args: promos: List[dict], check_upcoming_promos: bool
+
+    Returns: Filtered promotions.
+    """
+    filtered_promos = [e for e in promos if is_relevant_promo(
+        e, check_upcoming_promos)]
```
Comment (woctezuma) [inline src/filter_utils.py:65]: I would run [`black` formatter](https://github.com/psf/black) before any merge.

File: src/print_utils.py [hunk 7]
```diff
@@ -30,16 +65,25 @@ def to_discount_symbol(discount_type):
...
     for e in get_sorted_promos(filtered_promos, check_upcoming_promos):
         name = e["title"]
         meta_data = get_meta_data(e, check_upcoming_promos)
 
         start_date = trim_date(meta_data["startDate"])
         end_date = trim_date(meta_data["endDate"])
         discount = meta_data["discountSetting"]["discountPercentage"]
-        symbol = to_discount_symbol(meta_data["discountSetting"]["discountType"])
+        symbol = to_discount_symbol(
+            meta_data["discountSetting"]["discountType"])
```
Comment (woctezuma) [inline src/print_utils.py:83]: I would run [`black` formatter](https://github.com/psf/black) before any merge.

File: src/print_utils.py [hunk 8]
```diff
@@ -30,16 +65,25 @@ def to_discount_symbol(discount_type):
...
 
-        print(f"- [ {discount:2d}{symbol} off ] {start_date} -> {end_date} : {name}")
+        print(
+            f"- [ {discount:2d}{symbol} off ] {start_date} -> {end_date} : {name}")
```
Comment (woctezuma) [inline src/print_utils.py:86]: I would run [`black` formatter](https://github.com/psf/black) before any merge.

File: .gitignore [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+.spyproject
+
+# Rope project settings
+.ropeproject
+
+# mkdocs documentation
+/site
+
+# mypy
+.mypy_cache/
```
Comment (woctezuma) [inline .gitignore:101]: I would not include `.gitignore`.

File: src/aggregate_utils.py [hunk 10]
```diff
@@ -1,17 +1,39 @@
...
 
 def extract_upcoming_promotions(store_data):
+    """
+    Extract upcoming promotions from store data.
+
+    args: store_data: dict
+
+    Returns: List of upcoming promotions.
+    """
     promos = store_data["elements"]
```
Comment (woctezuma) [inline src/aggregate_utils.py:13]: This is a minor remark. Codacy Static Code Analysis mentions a minor issue with PEP 257 docstring style checker:
> Multi-line docstring summary should start at the first line (D212)

That is the style which is most common. However, it is true that [PEP 257](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings) allows both styles.
> The summary line may be on the same line as the opening quotes or on the next line.

**Comment by woctezuma** [comment]
Another minor remark: I see you edited the `KonScanner:main` branch in your fork. I believe the usual process for you would have been to create a new branch so that you could easily rebase if I edited my `main` branch.

File: src/print_utils.py [hunk 11]
```diff
@@ -3,25 +3,60 @@
...
     return f"{extract_day(date_str)} @ {extract_hour(date_str)}h"
 
 
 def extract_hour(date_str):
+    """
+    Extract hour from date string.
+
+    args: date_str: str
+
+    Returns: Hour: int
```
Comment (woctezuma) [inline src/print_utils.py:36]: The hour is actually returned as a string here.

File: src/aggregate_utils.py (hunk 1 above)
Comment (KonScanner) [reply src/aggregate_utils.py:1]: Sorry I was testing within repo, my bad! I shouldn't have rushed the merge req

File: src/aggregate_utils.py (hunk 4 above)
Comment (KonScanner) [reply src/aggregate_utils.py:50]: I was using `autopep8`, sorry, I should've checked!

File: src/aggregate_utils.py (hunk 5 above)
Comment (KonScanner) [reply src/aggregate_utils.py:83]: Feel free to add this func if you think it may be useful! :)

**Comment by woctezuma** [comment]
@KonScanner If you don't mind, I can implement some of your stuff, notably the doc strings and your regex search function. I will credit you in the code and commit.

It is just that I would take some of your pull request but not all of it, as I mentioned earlier, and I am not very used to managing pull requests. Plus, I don't want to bother you by asking changes, etc., especially as you work directly on `main`, so you cannot pull my `main` after I update it and then easily `rebase`.

I appreciate a lot your pull request! I was surprised that you would have gone through the code and implemented these changes so fast! Very fast reader and thinker! :D

**Comment by KonScanner** [comment]
Feel free to do so! I saw you implemented `typing` as well, which is something the docstrings aimed to do. There's no real need for docstrings now I think, your functions are well named and self-explanatory ;)

**Comment by woctezuma** [comment]
Yes, I thought about type hints thanks to your doc strings! ❤️

File: src/aggregate_utils.py (hunk 5 above)
Comment (woctezuma) [reply src/aggregate_utils.py:83]: Thanks. I will do! 👍

**Comment by woctezuma** [comment]
For info:

- the usage of `re.search()` was slightly wrong. I caught that with a unit test. :)



vs.

https://github.com/woctezuma/egs-15DaysofGames/blob/e8fc4871c0033c440c2ff9e3da266e7a430bacda/src/search_utils.py#L25-L29

- this might be the intended description, or this might be a wrong copy-paste.

**Comment by woctezuma** [comment]
Thank you for your contribution! 👍 

1) This pushed me to look into type hinting.
2) I appreciate the new search function.
3) The docstrings clarify that my previous function was only looking for the first match, compared to the new function.

**Pull request closed without merging**